pybaseball/datasources/fangraphs.py

```diff
@@ -1,44 +1,135 @@
+from abc import ABC
+from enum import Enum, unique
+from typing import Callable, Iterable, List, Optional, Type, Union
 
 import lxml
 import pandas as pd
 import requests
 
 from pybaseball.datahelpers import postprocessing
+from pybaseball.datahelpers.column_mapper import BattingStatsColumnMapper, GenericColumnMapper, ColumnListMapperFunction
+from pybaseball.datasources.html_table_processor import HTMLTableProcessor
+from pybaseball.enums.fangraphs import (FanGraphsStat, FanGraphsBattingStat, FanGraphsFieldingStat, FanGraphsLeague, FanGraphsMonth,
+                                        FanGraphsPitchingStat, FanGraphsPositions, FanGraphsStatTypes)
 
-ROOT_URL = 'https://www.fangraphs.com'
+_FG_LEADERS_URL = "/leaders.aspx"
 
+MIN_AGE = 0
+MAX_AGE = 100
 
-def get_fangraphs_tabular_data_from_html(html: Union[str, bytes], column_name_mapper: Callable = None, known_percentages: List[str] = []) -> pd.DataFrame:
-    xpath: str = '//table[@class="rgMasterTable"]'
-    html_dom = lxml.etree.HTML(html)
+class FanGraphsDataTable(ABC):
+    ROOT_URL: str = "https://www.fangraphs.com"
+    TABLE_CLASS: str = "rgMasterTable"
+    HEADINGS_XPATH: str = "({TABLE_XPATH}/thead//th[contains(@class, 'rgHeader')])[position()>1]/descendant-or-self::*/text()"
+    DATA_ROWS_XPATH: str = "({TABLE_XPATH}/tbody//tr)"
+    DATA_CELLS_XPATH: str = "td[position()>1]/descendant-or-self::*/text()"
+    QUERY_ENDPOINT: str = _FG_LEADERS_URL
+    STATS: FanGraphsStatTypes = FanGraphsStatTypes.NONE
+    DEFAULT_TYPES: FanGraphsStat = None
+    KNOWN_PERCENTAGES: List[str] = []
+    TEAM_DATA: bool = False
+    COLUMN_NAME_MAPPER: ColumnListMapperFunction = GenericColumnMapper().map_list
 
-    headings_xpath = f"({xpath}/thead//th[contains(@class, 'rgHeader')])[position()>1]/descendant-or-self::*/text()"
-    headings = html_dom.xpath(headings_xpath)
+    def __init__(self):
+        self.html_accessor = HTMLTableProcessor(
+            root_url=self.ROOT_URL,
+            headings_xpath=self.HEADINGS_XPATH,
+            data_cell_xpath=self.DATA_CELLS_XPATH,
+            data_rows_xpath=self.DATA_ROWS_XPATH,
+            table_class=self.TABLE_CLASS,
+        )
 
-    if column_name_mapper:
-        headings = [column_name_mapper(h) for h in headings]
+    def _postprocess(self, data: pd.DataFrame) -> pd.DataFrame:
+        return data
 
-    data_rows_xpath = f"({xpath}/tbody//tr)"
-    data_rows_dom = html_dom.xpath(data_rows_xpath)
-    data_rows = [
-        [
-            postprocessing.try_parse(y, headings[index], known_percentages=known_percentages)
-            for index, y in enumerate(x.xpath('td[position()>1]/descendant-or-self::*/text()'))
-        ]
-        for x in data_rows_dom
-    ]
+    def _validate(self, data: pd.DataFrame) -> pd.DataFrame:
+        return data
 
-    fg_data = pd.DataFrame(data_rows, columns=headings)
+    def fetch(self, start_season: int, end_season: Optional[int] = None, types: FanGraphsStat = None,
```

Comment:
how does using a non-default value for `stats` work? I tried calling it with `fg_batting_data(2019, types=FanGraphsBattingStat.BABIP.value)` and got a 500 error. I guess you need the `c` to be inserted before the list of integer stats codes?

Reply:
sorry, I meant non-default value for the argument, `types`

Reply:
Yeah, it seems you either have to specify the c (which I think stands for common stats), or it's components (Name, Team, Season). I should at least document that, and possibly fix the call to insert that. What do you think?

Reply:
And now that I'm thinking further I need to revamp some of this code. If the request is for a fixed set of columns from the user, this shouldn't be a stat coming in, but rather a list of them. Thanks for helping me see a blind spot there.

Reply:
a suggestion - what if as a first step we don't give the user a choice, you just get all the stats. the user is free to filter the columns of the resulting data frame if they want. in a subsequent PR we could fix it, but it seems potentially complicated and I feel like deferring and merging this useful PR first is better than either blocking this PR or rushing a non-optimal solution

Reply:
Yeah, I'll change my mind on this and side with @bdilday here

Reply:
I didn't see these until after I had done some tweaks here. I now have it working consistently.

For example:

```python
import pybaseball
from pybaseball.enums.fangraphs import FanGraphsBattingStat

pybaseball.datasources.fangraphs.fg_batting_data(2019, types=[FanGraphsBattingStat.AB])
```

Now returns:

|Season | Name | Team | AB
|-- | -- | -- | --
|2019 | Christian Yelich | Brewers | 489
|2019 | Mike Trout | Angels | 470
|2019 | Alex Bregman | Astros | 554
|2019 | Nelson Cruz | Twins | 454
|2019 | Cody Bellinger | Dodgers | 558
|... | ... | ... | ...
|2019 | Khris Davis | Athletics | 481
|2019 | Yolmer Sanchez | White Sox | 496
|2019 | Brandon Crawford | Giants | 500
|2019 | Mallex Smith | Mariners | 510
|2019 | Orlando Arcia | Brewers | 494

Let me know if this works.

Reply:
LGTM!

Reply:
Sorry, I _really_ hate to necromance this, but.
1. This is **really** good behavior to have.
2. We should allow:
```python
import pybaseball
pybaseball.fg_batting_data(2019, types=['AB'])
```

Reply:
Already in there that way :)

stat_columns accepts either a str or List[str]

```python
import pybaseball

pybaseball.datasources.fangraphs.fg_batting_data(2019, stat_columns=['AB'])
```

|Season | Name | Team | AB
|-- | -- | -- | --
|2019 | Christian Yelich | Brewers | 489
|2019 | Mike Trout | Angels | 470
|2019 | Alex Bregman | Astros | 554
|2019 | Nelson Cruz | Twins | 454
|2019 | Cody Bellinger | Dodgers | 558
|... | ... | ... | ...
|2019 | Khris Davis | Athletics | 481
|2019 | Yolmer Sanchez | White Sox | 496
|2019 | Brandon Crawford | Giants | 500
|2019 | Mallex Smith | Mariners | 510
|2019 | Orlando Arcia | Brewers | 494

135 rows × 4 columns